aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -739,103 +741,30 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
-void inline sparse_mask_out_cpu_kernel(
-    Tensor& r_values,
-    const Tensor& t,
-    const int64_t r_nnz,
-    const int64_t sparse_dim,
-    const Tensor& mask_indices) {
-  auto r_values_accessor = r_values.accessor<scalar_t, 1>();
-  auto mask_indices_accessor = mask_indices.accessor<int64_t, 2>();
-  scalar_t* t_ptr = t.data_ptr<scalar_t>();
-  auto t_strides = t.strides();
-
-  at::parallel_for(0, r_nnz, 1000, [&](int64_t start, int64_t end) {
-    for (const auto i : c10::irange(start, end)) {
-      int64_t idx = 0;
-      for (const auto d : c10::irange(sparse_dim)) {
-        idx += mask_indices_accessor[d][i] * t_strides[d];
-      }
-      r_values_accessor[i] = t_ptr[idx];
-    }
-  });
-}
-
-SparseTensor& sparse_mask_out_cpu(
-    SparseTensor& r,
-    const Tensor& t,
-    const SparseTensor& mask) {
-  TORCH_CHECK(mask.is_coalesced(), "sparse_mask: mask is uncoalesced");
+SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
   TORCH_CHECK(
       mask.sizes().equals(t.sizes()),
-      "sparse_mask: operands have incompatible sizes; self has size ",
+      "sparse_mask(): operands have incompatible sizes; self has size ",
       t.sizes(),
       " but mask has size ",
       mask.sizes());
-  AT_ASSERT(!t.is_cuda()); // we were supposed to have dispatched on this
-  TORCH_CHECK(
-      !r.is_cuda(), "sparse_mask: expected 'out' to be CPU, but got CUDA");
-  TORCH_CHECK(
-      !mask.is_cuda(), "sparse_mask: expected 'mask' to be CPU, but got CUDA");
-  at::resize_as_sparse_(r, mask);
-  if (mask._nnz() == 0) {
-    return r.zero_();
-  }
-  int64_t dim = t.dim();
-  int64_t sparse_dim = mask.sparse_dim();
-  Tensor mask_indices = mask._indices();
-  Tensor mask_values = mask._values();
-  Tensor r_values = at::empty(mask_values.sizes(), r._values().options());
-  alias_into_sparse(r, mask_indices.clone(), r_values);
-  r._coalesced_(mask.is_coalesced());
-  int64_t r_nnz = mask._nnz();
-  get_sparse_impl(r)->set_nnz_and_narrow(r_nnz);
-
-  if (t.numel() ==
-      0) { // if t is an empty tensor, there is no need to mask its elements
-    return r;
-  }
-
-  if (dim > sparse_dim) {
-    // Get a flattened sparse indices, similar to NOTE [ Flatten Sparse Indices
-    // ]. Keeping this implementation because it is faster than
-    // flatten_indices()
-    Tensor indices = at::zeros({mask._nnz()}, mask_indices.options());
-    for (const auto d : c10::irange(mask.sparse_dim())) {
-      indices.mul_(mask.size(d));
-      indices.add_(mask_indices.select(0, d));
-    }
 
-    std::vector<int64_t> view_size(1 + mask.dense_dim());
-    view_size[0] = -1;
-    for (const auto d : c10::irange(mask.dense_dim())) {
-      view_size[d + 1] = mask.size(mask.sparse_dim() + d);
-    }
+  if (!mask.is_coalesced()) {
+    TORCH_WARN_ONCE(
+        "sparse_mask(): uncoalesced mask may result in duplicate values in the result. "
+        "It is recommended to pass mask.coalesce() if such behavior is not expected.");
+  }
```

Comment:
I suggest moving this warning to `torch.sparse_mask` documentation because the described effect exists with other operations (mul, ...) as well that operate on uncoalesced tensors but do not warn. IIUC, `torch.sparse_mask(tensor, mask).coalesce() == torch.sparse_mask(tensor, mask.coalesce())` still holds here.